### PR TITLE
gadget: support validation of partial gadgets

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
This  PR introduces the new "partial" field in gadget.yaml and considers it when validating the gadget information.

### CODE REVIEW

File: gadget/gadget.go [hunk 1]
```diff
@@ -121,10 +121,25 @@ type Info struct {
...
+var validPartialValues = [...]PartialPiece{PartialSize, PartialFilesystem, PartialSchema, PartialStructure}
+
 // Volume defines the structure and content for the image to be written into a
 // block device.
 type Volume struct {
-	// Schema describes the schema used for the volume
+	// Partial is a list of features that are only only partially
+	// described in the gadget and that need to be filled by an
+	// installer. Schema describes the schema used for the volume.
+	Partial []PartialPiece `yaml:"partial" json:"partial"`
```
Comment (pedronis) [inline gadget/gadget.go:141]: these can be omitempty?

File: gadget/gadget.go [hunk 2]
```diff
@@ -121,10 +121,25 @@ type Info struct {
...
+
+var validPartialValues = [...]PartialPiece{PartialSize, PartialFilesystem, PartialSchema, PartialStructure}
+
 // Volume defines the structure and content for the image to be written into a
 // block device.
 type Volume struct {
-	// Schema describes the schema used for the volume
+	// Partial is a list of features that are only only partially
+	// described in the gadget and that need to be filled by an
+	// installer. Schema describes the schema used for the volume.
```
Comment (pedronis) [inline gadget/gadget.go:140]: why is Schema mentioned here?

File: gadget/gadget.go [hunk 3]
```diff
@@ -121,10 +121,25 @@ type Info struct {
 	KernelCmdline KernelCmdline `yaml:"kernel-cmdline"`
 }
 
+type PartialPiece string
```
Comment (pedronis) [inline gadget/gadget.go:124]: s/Piece/Property/ ?

File: gadget/gadget_test.go [hunk 4]
```diff
@@ -4448,3 +4448,179 @@ volumes:
...
+    structure:
+      - name: ubuntu-seed
+        filesystem: ext4
+        size: 500M
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        role: system-seed
+      # Space for some unknown structure in the middle is left around
+      - name: ubuntu-boot
+        filesystem: ext4
+        offset: 1000M
```
Comment (pedronis) [inline gadget/gadget_test.go:4616]: as by current definition things could end up on different disks we need a bit to think what it means to have offsets with partial: [structure]

**Review by pedronis — COMMENTED** [review]
did a first pass, it's a bit hard to tell of course if there are more places needing if partiall(prop) ?

File: gadget/gadget.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply gadget/gadget.go:124]: Good suggestion, I was struggling to find a good type name. Changed.

File: gadget/gadget.go (hunk 2 above)
Comment (alfonsosanchezbeato) [reply gadget/gadget.go:140]: Removed

File: gadget/gadget.go (hunk 1 above)
Comment (alfonsosanchezbeato) [reply gadget/gadget.go:141]: Added

File: gadget/gadget_test.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply gadget/gadget_test.go:4616]: As discussed, I've added a TODO as it is not clear yet what will a partial structure constrain.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the review, comments addressed. Related to more checks for partial properties, I've added a TODO for "structure".

**Review by pedronis — APPROVED** [review]
thanks

**Comment by alfonsosanchezbeato** [comment]
I have rebased now

File: gadget/gadget.go [hunk 5]
```diff
@@ -771,6 +823,28 @@ func orderStructuresByOffset(vss []VolumeStructure) []VolumeStructure {
...
+			if p == valid {
+				if foundCache[vpvIdx] == true {
+					return fmt.Errorf("partial value %q is repeated", p)
+				}
+				foundCache[vpvIdx] = true
+				found = true
+				break
+			}
+		}
+		if found == false {
```
Comment (mvo5) [inline gadget/gadget.go:840]: (super nitpick) I slightly prefer `if !found {` as it reads in my head as `if not found` which is slightly shorter than `if found is false' but this is probably the mother of all nitpicks :)

**Review by mvo5 — APPROVED** [review]
Thank you, this is looks great!

**Pull request merged**